src/transaction.rs

```diff
@@ -1,78 +1,360 @@
+    pub spend_slices: [SpendSlice<CP>; NUM_TX_SLICE],
+    pub output_slices: [OutputSlice<CP>; NUM_TX_SLICE],
 }
 
-impl<'a, CP: CircuitParameters> Transaction<'a, CP> {
-    pub fn new(
-        //max: usize,
-        actions: Vec<Action<CP>>,
-        spent_notes: Vec<(Note<CP>, Nullifier<CP>)>,
-        created_notes: Vec<(Note<CP>, EncryptedNote<CP::InnerCurve>)>,
-        vps: &'a Vec<ValidityPredicate<CP>>,
-    ) -> Self {
+pub struct ActionSlice<CP: CircuitParameters> {
+    pub action_proof: Proof<CP::CurveScalarField, CP::CurvePC>,
+    pub action_public: Action<CP>,
+}
+
+pub struct SpendSlice<CP: CircuitParameters> {
+    spend_addr_vp: VPCheck<CP>,
+    spend_token_vp: VPCheck<CP>,
+}
+
+pub struct OutputSlice<CP: CircuitParameters> {
+    output_addr_vp: VPCheck<CP>,
+    output_token_vp: VPCheck<CP>,
+}
+
+pub struct VPCheck<CP: CircuitParameters> {
+    // VP circuit size
+    pub vp_circuit_size: usize,
+    // VP proof
+    pub vp_proof: Proof<CP::CurveScalarField, CP::CurvePC>,
+    // The public inputs for vp proof
+    // TODO: maybe the PublicInputs<CP::CurveScalarField> should be serialized from Vec<CP::CurveScalarField>, like the action_public in ActionSlice?
+    pub vp_public_inputs: PublicInputs<CP::CurveScalarField>,
+    // The vk of vp is blinded partially, blinded parts can be constructed from blinded_vp_public_inputs.
+    // The unblinded_vp_vk is the unblinded parts.
+    // TODO:
+    pub unblinded_vp_vk: VerifierKey<CP::CurveScalarField, CP::CurvePC>,
+    // blind vp proof
+    pub blind_vp_proof: Proof<CP::CurveBaseField, CP::OuterCurvePC>,
+    // The public inputs for blind vp proof
+    pub blinded_vp_public_inputs: PublicInputs<CP::CurveBaseField>,
+    // TODO:
+    // pub vp_com
+    // pub vp_param
+    // pub vp_memo
+}
+
+impl<CP: CircuitParameters> SpendSlice<CP> {
+    pub fn new(spend_addr_vp: VPCheck<CP>, spend_token_vp: VPCheck<CP>) -> Self {
         Self {
-            //max,
-            actions,
-            spent_notes,
-            created_notes,
-            vps,
+            spend_addr_vp,
+            spend_token_vp,
         }
     }
+}
 
-    fn check(&self) {
-        //1. action check
-
-        //2. verify validity predicates;
-        //2.1 todo: update to verification of blinded vps
-        //2.2 todo: add blinding circuit check
-        for vp in self.vps {
-            vp.verify()
+impl<CP: CircuitParameters> OutputSlice<CP> {
+    pub fn new(output_addr_vp: VPCheck<CP>, output_token_vp: VPCheck<CP>) -> Self {
+        Self {
+            output_addr_vp,
+            output_token_vp,
         }
     }
+}
+
+impl<CP: CircuitParameters> ActionSlice<CP> {
+    pub fn build(
+        action_public: Action<CP>,
+        action_circuit: &mut ActionCircuit<CP>,
+    ) -> Result<Self, TaigaError> {
+        let setup = CP::get_pc_setup_params(ACTION_CIRCUIT_SIZE);
+        // Compile the circuit
+        let pk_p = CP::get_action_pk();
```

Comment:
(useless comment)
For me, "pk" means "prover key" and "vk" means "verifier key" so I guess we should remvoe `_p`.